### PR TITLE
Modernize the Ray Tune tutorial

### PR DESCRIPTION
**Pull request by crypdick**
This is the first of several PRs against the Pytorch Tutorials. This PR unstales the Ray Tune tutorial:

- bumps Ray version from 2.7.2 -> 2.52.1
- update the tutorial with the latest code from the Pytorch CIFAR10 tutorial which it is based off of
- update to use the latest Ray Tune APIs
- fixed the norm values for CIFAR10
- rewrote most of the prose
- added new sections for sampling, observability, conclusion, additional links
- updated tags in the doc indices. Will need approval on the new `Ray-Distributed` tag.

Opening this as a draft PR so that I can get feedback from the Ray Tune core devs first

cc @subramen @albanD @jbschlosser @pcmoritz @robertnishihara @matthewdeng @richardliaw

### CODE REVIEW

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 1]
```diff
@@ -436,77 +479,118 @@ def main(num_samples=10, max_num_epochs=10, gpus_per_trial=2):
...
         print("Best trial test set accuracy: {}".format(test_acc))
 
 
 if __name__ == "__main__":
-    # You can change the number of GPUs per trial here:
-    main(num_samples=10, max_num_epochs=10, gpus_per_trial=0)
-
+    # Set the number of trials, epochs, and GPUs per trial here:
+    # The following configuration is for a quick run (1 trial, 1 epoch, CPU only) for demonstration purposes.
+    main(num_trials=1, max_num_epochs=1, gpus_per_trial=0)
```
Comment (crypdick) [inline beginner_source/hyperparameter_tuning_tutorial.py]: TODO: add GPUs back when I am done

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 2]
```diff
@@ -50,33 +46,22 @@
...
-# We wrap the data loaders in their own function and pass a global data directory.
-# This way we can share a data directory between different trials.
-
+# Wrap the data loaders in a constructor function. Pass a global data
+# directory here to reuse the dataset across different trials.
 
 def load_data(data_dir="./data"):
     transform = transforms.Compose(
-        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+        [transforms.ToTensor(), transforms.Normalize((0.4914, 0.48216, 0.44653), (0.2022, 0.19932, 0.20086))]
```
Comment (crypdick) [inline beginner_source/hyperparameter_tuning_tutorial.py:85]: I wrote a blog post about this: https://www.ricardodecal.com/guides/use-these-normalization-values-for-torchvision-datasets/

**Comment by crypdick** [comment]
cc @xinyuangui2

File: beginner_source/hyperparameter_tuning_tutorial.py (hunk 2 above)
Comment (xinyuangui2) [reply beginner_source/hyperparameter_tuning_tutorial.py:85]: Let's comment in the code so users can understand those magic numbers.

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 3]
```diff
@@ -117,119 +102,133 @@ def forward(self, x):
...
-#         device = "cuda:0"
-#         if torch.cuda.device_count() > 1:
-#             net = nn.DataParallel(net)
-#     net.to(device)
+#    device = "cpu"
+#    if torch.cuda.is_available():
+#        device = "cuda:0"
+#        if torch.cuda.device_count() > 1:
+#            net = nn.DataParallel(net)
+#    net.to(device)
```
Comment (xinyuangui2) [inline beginner_source/hyperparameter_tuning_tutorial.py]: If you use nn.DataParallel, you should move the model to CUDA before wrapping it, or else PyTorch will warn and may move things incorrectly.

can change to 
```
if torch.cuda.is_available():
    net = net.to("cuda")      # move to GPU first
    if torch.cuda.device_count() > 1:
        net = nn.DataParallel(net)
else:
    net = net.to("cpu")
```

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 4]
```diff
@@ -117,119 +102,133 @@ def forward(self, x):
...
+# remains standard PyTorch.
 #
 # Full training function
 # ~~~~~~~~~~~~~~~~~~~~~~
 #
 # The full code example looks like this:
 
-
 def train_cifar(config, data_dir=None):
     net = Net(config["l1"], config["l2"])
```
Comment (xinyuangui2) [inline beginner_source/hyperparameter_tuning_tutorial.py:170]: Remember to update this device moving logic.

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 5]
```diff
@@ -89,18 +76,16 @@ def load_data(data_dir="./data"):
...
-# ---------------------------
-# We can only tune those parameters that are configurable.
-# In this example, we can specify
-# the layer sizes of the fully connected layers:
-
+# Configure the hyperparameters
+# -----------------------------
+#
+# In this example, we specify the layer sizes of the fully connected
+# layers.
```
Comment (xinyuangui2) [inline beginner_source/hyperparameter_tuning_tutorial.py]: Before the details, let's tell users the parameters that we tune: linear layer size and learning rate.

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 6]
```diff
@@ -117,119 +102,133 @@ def forward(self, x):
...
-#     optimizer = optim.SGD(net.parameters(), lr=config["lr"], momentum=0.9)
+#    optimizer = optim.SGD(net.parameters(), lr=config["lr"], momentum=0.9)
 #
-# We also split the training data into a training and validation subset. We thus train on
-# 80% of the data and calculate the validation loss on the remaining 20%. The batch sizes
-# with which we iterate through the training and test sets are configurable as well.
+# We also split the training data into a training and validation subset.
+# We thus train on 80% of the data and calculate the validation loss on
+# the remaining 20%. The batch sizes with which we iterate through the
+# training and test sets are configurable as well.
```
Comment (xinyuangui2) [inline beginner_source/hyperparameter_tuning_tutorial.py]:
```suggestion
# training and test sets are configurable through ray tune.
```

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 7]
```diff
@@ -354,131 +365,119 @@ def test_accuracy(net, device="cpu"):
...
-# ``tune.with_resources``:
+# Tell Ray Tune what resources to allocate for each trial by passing a
+# ``resources`` dictionary to ``tune.with_resources``:
 #
 # .. code-block:: python
 #
 #    tune.with_resources(
 #        partial(train_cifar, data_dir=data_dir),
-#        resources={"cpu": cpus_per_trial, "gpu": gpus_per_trial}
+#        resources={"cpu": 2, "gpu": gpus_per_trial}
```
Comment (xinyuangui2) [inline beginner_source/hyperparameter_tuning_tutorial.py]: nit: why do we hard code here?

Comment (crypdick) [reply beginner_source/hyperparameter_tuning_tutorial.py]: nice catch, forgot to revert after testing

**Comment by crypdick** [comment]
@svekars Ready for review from the PyTorch side :)

**Comment by crypdick** [comment]
Happy new year @svekars . Gentle bump on the review on this PR. This one is soft blocking other contributions because of the Ray version bump.

**Comment by svekars** [comment]
Thanks for the reminder! Will review today.

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 8]
```diff
@@ -1,44 +1,43 @@
...
 
-1. wrap data loading and training in functions,
-2. make some network parameters configurable,
-3. add checkpointing (optional),
-4. and define the search space for the model tuning
+Setup
+-----
 
-|
+To run this tutorial, install the dependencies:
```
Comment (svekars) [inline beginner_source/hyperparameter_tuning_tutorial.py]:
```suggestion
To run this tutorial, install the following dependencies:
```

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 9]
```diff
@@ -1,44 +1,43 @@
...
-a network layer size can have a dramatic impact on your model performance.
+**Author:** `Ricardo Decal <https://github.com/crypdick>`__
 
-Fortunately, there are tools that help with finding the best combination of parameters.
-`Ray Tune <https://docs.ray.io/en/latest/tune.html>`_ is an industry standard tool for
-distributed hyperparameter tuning. Ray Tune includes the latest hyperparameter search
-algorithms, integrates with various analysis libraries, and natively
-supports distributed training through `Ray's distributed machine learning engine
-<https://ray.io/>`_.
+This tutorial shows how to integrate Ray Tune into your PyTorch training
```
Comment (svekars) [inline beginner_source/hyperparameter_tuning_tutorial.py:7]: After a short intro, can you add two sections (or cards like [this](https://github.com/pytorch/tutorials/blob/main/beginner_source/template_tutorial.py#L9)):
- Prerequisites (minimum PyTorch version, ray tune version, GPU/needed or not, etc)
- What you will learn (what will the user learn after completing this tutorial)

**Review by svekars — COMMENTED** [review]
I also see that now it takes 32 minutes to [build](https://docs-preview.pytorch.org/pytorch/tutorials/3694/beginner/hyperparameter_tuning_tutorial.html#further-reading) this tutorial while the [current version ](https://docs.pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html) takes 12 minutes. Can we reduce the time of the build. We could try to put on a different [worker](https://github.com/pytorch/tutorials/blob/main/.jenkins/metadata.json) and see if it helps.

File: beginner_source/hyperparameter_tuning_tutorial.py (hunk 9 above)
Comment (crypdick) [reply beginner_source/hyperparameter_tuning_tutorial.py:7]: @svekars done 👍

**Comment by crypdick** [comment]
@svekars I made the following changes to speed up the tutorial:

- I reverted the number of data loader workers from 2 back to 8
- I made a GPU available to the tutorial by changing the CI worker to linux.g5.4xlarge.nvidia.gpu

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 10]
```diff
@@ -355,158 +380,232 @@ def test_accuracy(net, device="cpu"):
...
-# You just have to make sure that the models still fit in the GPU memory.
+# For example, if you are running this experiment on a cluster of 20
+# machines, each with 8 GPUs, you can set ``gpus_per_trial = 0.5`` to
+# schedule two concurrent trials per GPU. This configuration runs 320
+# trials in parallel across the cluster.
 #
-# After training the models, we will find the best performing one and load the trained
-# network from the checkpoint file. We then obtain the test set accuracy and report
-# everything by printing.
+#    **Note**: To run this tutorial without GPUs, set ``gpus_per_trial=0``
```
Comment (svekars) [inline beginner_source/hyperparameter_tuning_tutorial.py]: can we please use the note directive
```suggestion
# .. note::
#    To run this tutorial without GPUs, set ``gpus_per_trial=0``
```

File: beginner_source/hyperparameter_tuning_tutorial.py [hunk 11]
```diff
@@ -355,158 +380,232 @@ def test_accuracy(net, device="cpu"):
...
+# tools <https://docs.ray.io/en/latest/ray-observability/index.html>`__
+# that let you attach a debugger to running trials across the cluster.
+#
+# Conclusion
+# ==========
+#
+# In this tutorial, you learned how to tune the hyperparameters of a
+# PyTorch model using Ray Tune. You saw how to integrate Ray Tune into
+# your PyTorch training loop, define a search space for your
+# hyperparameters, use an efficient scheduler like ASHAScheduler to
```
Comment (svekars) [inline beginner_source/hyperparameter_tuning_tutorial.py]:
```suggestion
# hyperparameters, use an efficient scheduler like ``ASHAScheduler`` to
```

**Review by svekars — APPROVED** [review]
LGTM - I added a couple of very minor formatting comments! Thank you for adjusting the execution time and updating the tutorial!
